### ISSUE TITLE
Cannot return PropertyDescriptor from property decorator in Visual Studio

### ISSUE THREAD

**Issue by JohnWeisz** [Microsoft/TypeScript#11661]
**TypeScript Version:**  2.0.2.0
**Visual Studio Version:** Community 2015 / 14.0.25425.01 Update 3

**Code**

``` ts

function propDeco(target: Object | Function, propKey: string | symbol): PropertyDescriptor {
    return {
        get: () => null
    };
}

class Person {
    @propDeco
    name: string;
}
```

**Expected behavior:**

According to [Decorators](https://www.typescriptlang.org/docs/handbook/decorators.html) (section _Property Decorators_), no errors should be reported.

**Actual behavior:** 

The following error is reported in Visual Studio, on the line of `@propDeco`:

> The return type of a property decorator function must be either 'void' or 'any'.
> 
> > Unable to resolve signature of property decorator when called as an expression.

If the _Do not emit outputs if any errors are reported_ box is unchecked, the compilation will be done, but the error will still be displayed in the editor. This leads me to believe this is a problem with the TypeScript distribution that is shipped with Visual Studio itself.

---

**Update**

On line `1276` of `Common7\IDE\CommonExtensions\Microsoft\TypeScript\lib.d.ts` is the following:

``` ts
declare type PropertyDecorator = (target: Object, propertyKey: string | symbol) => void;
```

I believe this definition is incorrect.

**Comment by mhegazy** [comment]
> I believe this definition is incorrect.

No. this is intentional. you can only change the descriptors for methods, and not for members. this is done in anticipation for the property declaration proposal in future versions of ECMAScript.

The main issue is members do not live on the prototype, but on the object, where methods do. so setting a `value` on the descriptor would mean it has to execute every time the constructor is invoked and not at class construction time.

if you want to defined a getter/setter on the protoype, use `Object.defineProperty` explicitly during the decoration process.
